### PR TITLE
Api free and reserve ticket

### PR DESCRIPTION
**Pull request by xFLy3r**
Doctrine was updated cause it has problems with php 7.1. See [link](https://github.com/doctrine/doctrine2/issues/6188).

### CODE REVIEW

File: src/AppBundle/Repository/CustomerOrderRepository.php [hunk 1]
```diff
@@ -2,7 +2,29 @@
...
 class CustomerOrderRepository extends AbstractRepository
 {
-
+    /**
+     * Method returns opened customer order
+     *
+     * @param Customer $param
+     * @return CustomerOrder|null
+     */
+    public function findLastOpenOrder(Customer $param)
```
Comment (sobchenko) [inline src/AppBundle/Repository/CustomerOrderRepository.php]: in my opinion findLastOpenOrder is not necessary here use findOneBy

File: src/AppBundle/Entity/Ticket.php [hunk 2]
```diff
@@ -95,8 +95,7 @@ class Ticket
     /**
      * @var CustomerOrder
      *
-     * @ORM\ManyToOne(targetEntity="CustomerOrder", inversedBy="tickets")
-     * @ORM\Column(name="customer_order_id", type="integer", nullable=true)
+     * @ORM\ManyToOne(targetEntity="CustomerOrder", inversedBy="tickets", fetch="EAGER")
      */
```
Comment (sobchenko) [inline src/AppBundle/Entity/Ticket.php]: Leave line 99 it is important here!!!

File: src/AppBundle/Controller/TicketsController.php [hunk 3]
```diff
@@ -48,13 +51,18 @@ public function freeAction(Ticket $id)
      * @Patch(requirements={"id" = "[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}"})
      * @ParamConverter("id", class="AppBundle:Ticket")
      */
-    public function reserveAction(Ticket $id)
+    public function reserveAction(Ticket $id, Request $request)
     {
         //This done not in right way (Ticket $ticket) to have RESTfully looking route: /tickets/{id}
         $ticket = $id;
 
+        $apiKey =  $request->headers->get('API-Key-Token');
```
Comment (spolischook) [inline src/AppBundle/Controller/TicketsController.php]: use 
```php
$this->getUser()
```

File: src/AppBundle/Services/OrderManager.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+use AppBundle\Entity\Ticket;
+use AppBundle\Entity\UserOrder;
+use AppBundle\Repository\UserOrderRepository;
+use Symfony\Bridge\Doctrine\RegistryInterface;
+use Symfony\Component\HttpKernel\Exception\AccessDeniedHttpException;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+
+class OrderManager
+{
+    protected $doctrine;
```
Comment (spolischook) [inline src/AppBundle/Services/OrderManager.php]: add docblock

File: src/AppBundle/Services/OrderManager.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+    protected $tokenStorage;
```
Comment (spolischook) [inline src/AppBundle/Services/OrderManager.php]: add docblock

File: src/AppBundle/Services/OrderManager.php [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    public function __construct(RegistryInterface $doctrine, TokenStorageInterface $tokenStorage)
+    {
+        $this->doctrine = $doctrine;
+        $this->tokenStorage = $tokenStorage;
+    }
+
+    public function addOrderToTicket(Ticket $ticket)
```
Comment (spolischook) [inline src/AppBundle/Services/OrderManager.php]: add return value type hinting

File: src/AppBundle/Services/OrderManager.php [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        $order = $this->getCustomerOrder();
+        $ticket->setUserOrder($order);
+    }
+
+    public function removeOrderFromTicket(Ticket $ticket)
```
Comment (spolischook) [inline src/AppBundle/Services/OrderManager.php]: add return value type hinting

File: src/AppBundle/Controller/TicketsController.php [hunk 8]
```diff
@@ -54,7 +55,11 @@ public function reserveAction(Ticket $id)
         $ticket = $id;
 
         $em = $this->getDoctrine()->getManager();
+        if ($ticket->getStatus() === Ticket::STATUS_BOOKED) {
+            throw new TicketStatusConflictException('Ticket is already booked');
```
Comment (spolischook) [inline src/AppBundle/Controller/TicketsController.php]: make this check and throw this exception in setStatus method of Ticket

File: src/AppBundle/EventListener/TicketStatusListener.php [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        
+        $oldStatus = $args->getOldValue('status');
+        $newStatus = $args->getNewValue('status');
+
+        if (!in_array($newStatus, Ticket::getStatuses())) {
+            throw new \InvalidArgumentException("Invalid ticket status");
+        }
+
+        if ($oldStatus === Ticket::STATUS_PAID) {
+            throw new TicketStatusConflictException("Invalid status. Ticket already paid.");
```
Comment (spolischook) [inline src/AppBundle/EventListener/TicketStatusListener.php]: move this to setStatus method of Ticket

File: src/AppBundle/Services/OrderManager.php [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    public function addOrderToTicket(Ticket $ticket): void
+    {
+        $order = $this->getCustomerOrder();
+        $ticket->setUserOrder($order);
+    }
+
+    /**
+     * @param Ticket $ticket
+     */
+    public function removeOrderFromTicket(Ticket $ticket): void
```
Comment (spolischook) [inline src/AppBundle/Services/OrderManager.php]: this semantically not correct. It should be ```removeTicketFromOrder```

File: src/AppBundle/Services/OrderManager.php [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        $em = $this->doctrine->getEntityManager();
+        $user = $this->tokenStorage->getToken()->getUser();
+        /** @var UserOrder $order */
+        $order = $em->getRepository('AppBundle:UserOrder')->findLastOpenOrder($user);
+        if ($ticket->getUserOrder() !== $order) {
+            throw new AccessDeniedHttpException('You cannot change status for this ticket');
```
Comment (spolischook) [inline src/AppBundle/Services/OrderManager.php]: throw http exception only in Controller. This service should used in other contexts that api

File: src/AppBundle/Services/OrderManager.php [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    /**
+     * @param Ticket $ticket
+     */
+    public function removeOrderFromTicket(Ticket $ticket): void
+    {
+        $em = $this->doctrine->getEntityManager();
+        $user = $this->tokenStorage->getToken()->getUser();
+        /** @var UserOrder $order */
+        $order = $em->getRepository('AppBundle:UserOrder')->findLastOpenOrder($user);
+        if ($ticket->getUserOrder() !== $order) {
```
Comment (spolischook) [inline src/AppBundle/Services/OrderManager.php]: This is not correct. Use security mechanism to determine if user has permissions to remove ticket from order

File: src/AppBundle/Services/OrderManager.php [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+            throw new AccessDeniedHttpException('You cannot change status for this ticket');
+        }
+
+        $ticket->setUserOrder(null);
+        $ticket->setStatus(Ticket::STATUS_FREE);
```
Comment (spolischook) [inline src/AppBundle/Services/OrderManager.php]: set status in setUserOrder method.

File: src/AppBundle/Services/OrderManager.php (hunk 12 above)
Comment (spolischook) [inline src/AppBundle/Services/OrderManager.php]: http://symfony.com/doc/current/security/voters.html

File: src/AppBundle/Entity/Ticket.php [hunk 14]
```diff
@@ -213,6 +214,13 @@ public function getStatus(): string
...
     public function setStatus($status)
     {
+        if ($this->getStatus() === Ticket::STATUS_PAID) {
+            throw new TicketStatusConflictException("Invalid status. Ticket already paid.");
+        }
+        if (($this->getStatus()  === Ticket::STATUS_BOOKED) && ($status === Ticket::STATUS_BOOKED)) {
```
Comment (spolischook) [inline src/AppBundle/Entity/Ticket.php]: what the hell here?
```php
if ($this->getStatus() === $status) {
  // nothing todo
  return $this;
}

File: src/AppBundle/Entity/Ticket.php [hunk 15]
```diff
@@ -213,6 +214,13 @@ public function getStatus(): string
      */
     public function setStatus($status)
     {
+        if ($this->getStatus() === Ticket::STATUS_PAID) {
```
Comment (spolischook) [inline src/AppBundle/Entity/Ticket.php]:
```php
if (!in_array($status, $this->getStatuses()) {
  throw new TicketStatusValidationException()
}

File: src/AppBundle/Entity/Ticket.php [hunk 16]
```diff
@@ -246,4 +254,31 @@ public function getSeriesNumber(): string
...
     }
+
+    /**
+     * Set customerOrder
+     *
+     * @param UserOrder $userOrder
+     *
+     * @return Ticket
+     */
+    public function setUserOrder(UserOrder $userOrder = null)
```
Comment (spolischook) [inline src/AppBundle/Entity/Ticket.php]: just add method removeOrder() and do what you need todo. Remove default ```null``` valued from $userOrder parameter

File: src/AppBundle/Entity/Ticket.php [hunk 17]
```diff
@@ -246,4 +254,31 @@ public function getSeriesNumber(): string
...
+    {
+        $status = $userOrder ? Ticket::STATUS_BOOKED : Ticket::STATUS_FREE;
+        $this->setStatus($status);
```
Comment (spolischook) [inline src/AppBundle/Entity/Ticket.php]: or just deny set order if it not null - it's the same might be.

File: src/AppBundle/Entity/Ticket.php [hunk 18]
```diff
@@ -213,6 +214,18 @@ public function getStatus(): string
      */
     public function setStatus($status)
     {
+        if (!in_array($status, $this->getStatuses())) {
```
Comment (spolischook) [inline src/AppBundle/Entity/Ticket.php]: this logic should be covered by unit tests, is it?

Comment (xFLy3r) [inline src/AppBundle/Entity/Ticket.php]: @spolischook, done.

**Comment by spolischook** [comment]
@xFLy3r  thanks for your contribution! 
Hope you are well!
Unfortunately we decide use third party service for sell tickets, so I close this PR

**Pull request closed without merging**